modules/sdk-coin-ada/src/lib/transactionBuilder.ts

```diff
@@ -102,21 +121,30 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
         )
       );
     });
-    const outputs = CardanoWasm.TransactionOutputs.new();
+    let outputs = CardanoWasm.TransactionOutputs.new();
+    let totalAmountToSend = CardanoWasm.BigNum.zero();
     this._transactionOutputs.forEach((output) => {
+      const amount = CardanoWasm.BigNum.from_str(output.amount);
       outputs.add(
         CardanoWasm.TransactionOutput.new(
           CardanoWasm.Address.from_bech32(output.address),
-          CardanoWasm.Value.new(CardanoWasm.BigNum.from_str(output.amount))
+          CardanoWasm.Value.new(amount)
         )
       );
+      totalAmountToSend = totalAmountToSend.checked_add(amount);
     });
+
+    // add extra output for the change
+    const changeAddress = CardanoWasm.Address.from_bech32(this._changeAddress);
```

Comment:
we needs to check if _changeAddress is undefined. When we reconstruct a txn from serialized data, the change address will not be set, and we can directly use output to build, and we can calculate fee based on the size of serialized data